### ISSUE TITLE
Add an exception for responses that are prematurely cut off due to their size.

### ISSUE THREAD

**Issue by andras-ferencz** [aspnet/KestrelHttpServer#2879]
This is a "feature" that has plagued me for months, until I realized there was a very easy solution to it ( [https://stackoverflow.com/questions/16436533/how-to-modify-the-default-allowed-response-size-settings-for-a-web-api-applicati] ).

Unless configured, there seems to be a (configurable) upper limit on the response size for JSON strings and responses in general - instead of throwing returning 500 and an exception that tells us the response is too large for the server to handle, it trims the response to the limit and sends it, leading to malformed JSON strings and 〈capslock〉developer frustration〈/capslock〉. 

I feel that an  exception (maybe can be toggled on/off) should be implemented here that at least tells us the server encountered an error instead of sending half of a response that inevitably breaks the client.

🐊

**Comment by Tratcher** [comment]
Kestrel doesn't limit response sizes, only request sizes. What do your application logs say?

**Comment by andras-ferencz** [comment]
Yeah, I think I have miscategorized this: it seems that the JSON Serializer  itself has a limit - but the configuration to fix the issue is in web.config (which is Kestrel, right?) 
```
       <system.web.extensions>
           <scripting>
               <webServices>
                   <jsonSerialization maxJsonLength="50000000"/>
               </webServices>
           </scripting>
       </system.web.extensions>

``` 

leading to my confusion of where to actually report this issue. It's strange that the the default upper limit is so small to begin with.

It actually doesn't say anything! No exceptions get raised here, it just cuts off my JSON abruptly without rhyme or reason once it hits that limit.

**Comment by Tratcher** [comment]
No, web.config.extensions is unrelated to Kestrel. I don't any mention that it applies to ASP.NET Core apps at all. @rynowak is there a matching setting for this in MVC?

**Comment by rynowak** [comment]
I'm not aware of any setting in MVC that does this. You'd have to do it inside the stream/writer. One way to accomplish this would be with a middleware/filter that replaces the body. You'd need to buffer if you want a 500 because otherwise you'd start writing to the response really early.

**Comment by muratg** [comment]
@andras-ferencz Did you figure out what the root cause was?

**Comment by samsosa** [comment]
Maybe the JavaScriptSerializer.MaxJsonLength Property?

Namespace:
System.Web.Script.Serialization
Assembly:
System.Web.Extensions.dll

From [MSDN](https://docs.microsoft.com/en-us/dotnet/api/system.web.script.serialization.javascriptserializer.maxjsonlength?redirectedfrom=MSDN&view=netframework-4.7.2#System_Web_Script_Serialization_JavaScriptSerializer_MaxJsonLength)

> Use the **jsonSerialization** element of the configuration file to set the maximum length for the internal serializer instance that is used by the asynchronous communication layer.

> The default is 2097152 characters, which is equivalent to 4 MB of Unicode string data.
